pandas/types/cast.py

```diff
@@ -983,3 +984,60 @@ def find_common_type(types):
             return np.object
 
     return np.find_common_type(types, [])
+
+
+def try_cast(arr, take_fast_path, data, dtype=None, copy=False,
+             raise_cast_failure=False):
+    """ try casting the array using dtype. Used in _sanitize_array procedure
+    at Series initialization
+    """
+    from pandas.core.categorical import Categorical, CategoricalAccessor
+
+    # perf shortcut as this is the most common case
+    if take_fast_path:
+        if maybe_castable(arr) and not copy and dtype is None:
+            return arr
+
+    try:
+        subarr = maybe_cast_to_datetime(arr, dtype)
+        if not is_extension_type(subarr):
+            subarr = np.array(subarr, dtype=dtype, copy=copy)
+
+            if dtype is not None:
+                d2 = data
+                if not is_list_like(d2):
+                    d2 = [data]
+                # Raises if coercion from unsigned to signed with neg data
+                if is_unsigned_integer_dtype(dtype) and (d2 < 0).any():
+                    raise OverflowError(
+                        "Trying to coerce negative values to negative "
+                        "integers")
+
+                # Raises if coercion from float to int with fraction data
+                # is_float_dtype(type(d2)) not working
+                if copy or not is_integer_dtype(data):
+                    assert_safe_casting(data, subarr)
+
+                if any([is_float_dtype(type(x)) for x in
+                        d2]) and is_integer_dtype(dtype):
+                    raise OverflowError(
+                        "Trying to coerce float values to integers")
+
+    except (ValueError, TypeError):
+        if is_categorical_dtype(dtype):
+            subarr = Categorical(arr)
+        elif dtype is not None and raise_cast_failure:
+            raise
+        else:
+            subarr = np.array(arr, dtype=object, copy=copy)
+    return subarr
+
+
+def assert_safe_casting(data, subarr):
+    """
+    Ensure incoming data can be represented as ints.
+    """
```

Comment:
this routine is all over the place (in Index), will want to simply call this one.